src/operator/tensor/init_op.h

```diff
@@ -164,19 +164,38 @@ inline bool InitStorageType(const nnvm::NodeAttrs& attrs,
   return true;
 }
 
+/*! \brief Fill output with a scalar integer value */
 template<typename xpu, int value>
 void FillCompute(const nnvm::NodeAttrs& attrs,
                  const OpContext& ctx,
                  const std::vector<TBlob>& inputs,
                  const std::vector<OpReqType>& req,
                  const std::vector<TBlob>& outputs) {
-  using namespace mshadow;
-  using namespace mshadow::expr;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
-    Tensor<xpu, 1, DType> out = outputs[0].FlatTo1D<xpu, DType>(s);
-    ASSIGN_DISPATCH(out, req[0], scalar<DType>(value));
-  });
+  if (req[0] != kNullOp) {
+    mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+    MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+      mxnet_op::Kernel<mxnet_op::set_to<value>, xpu>::Launch(s,
+                                                             outputs[0].Size(),
+                                                             outputs[0].dptr<DType>());
+    });
+  }
+}
+
+/*! \brief Fast CPU fill-zero version using memset */
+template<>
+inline void FillCompute<cpu, 0>(const nnvm::NodeAttrs& attrs,
+                                const OpContext& ctx,
+                                const std::vector<TBlob>& inputs,
+                                const std::vector<OpReqType>& req,
+                                const std::vector<TBlob>& outputs) {
+  if (req[0] != kNullOp) {
+    const size_t size = outputs[0].Size();
+    if (size) {
+      MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+        memset(outputs[0].dptr<DType>(), 0, size * sizeof(DType));
```

Comment:
`outputs[0].dptr_` is more efficient here than `outputs[0].dptr<DType>()`.
Question: How much faster is this compared to the original implementation of filling up an `TBlob`?

Reply:
a lot faster. assembly for memset looks something like:
; for 32 bit, load each register byte with destination value
mov eax, <val>
shl eax, 8
or al, ah
shl eax, 8
or al, ah
shl eax, 8
mov edi, <pointer>
mov ecx, <size>
shr ecx, 2
rep stosd  ; clock cycle coun, size * t: https://web.itu.edu.tr/kesgin/mul06/intel/instr/stos.html
mov ecx, <size>
and ecx, 3
rep stosb

This is done without any change in code execution path (jmp, jle, jge, jz, etc) or compares
It is also very cache-friendly since it is perfectly sequential.